Obsidian.IO/ProtocolWriter.cs

```diff
@@ -0,0 +1,418 @@
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                buffer[index++] = 0;
+                return;
+            }
+
+            var byteLength = Encoding.UTF8.GetByteCount(value);
+            WriteVarInt(byteLength);
+            EnsureCapacity(byteLength);
+            Encoding.UTF8.GetBytes(value, 0, value.Length, buffer, index);
+            index += byteLength;
+        }
+        
+        /// <summary>
+        /// Writes a <see cref="string"/> prefixed by <see cref="ushort"/>.
+        /// </summary>
+        public void WriteUShortString(string? value)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                buffer[index++] = 0;
+                return;
+            }
+
+            var byteLength = Encoding.UTF8.GetByteCount(value);
+            if (byteLength > ushort.MaxValue)
+                throw new ArgumentOutOfRangeException(nameof(value), "String is too big");
+            WriteUInt16((ushort) byteLength);
+            EnsureCapacity(byteLength);
+            Encoding.UTF8.GetBytes(value, 0, value.Length, buffer, index);
+            index += byteLength;
+        }
+
+        public unsafe void WriteGuid(Guid value)
+        {
+            EnsureCapacity(16);
+            var ptr = (byte*)&value;
+            fixed (byte* bptr = buffer)
+            {
+                System.Buffer.MemoryCopy(ptr, bptr + index, buffer.Length, 16);
+            }
+            index += 16;
+        }
+        #endregion
+
+        #region Writing arrays
+        public unsafe void WriteSpan(ReadOnlySpan<byte> span)
+        {
+            EnsureCapacity(span.Length);
+            fixed (byte* spanPtr = span)
+            {
+                fixed (byte* arrPtr = buffer)
+                {
+                    System.Buffer.MemoryCopy(spanPtr, arrPtr + index, buffer.Length - index, span.Length);
+                }
+            }
+            index += span.Length;
+        }
+
+        public void WriteInt64Span(ReadOnlySpan<long> value)
+        {
+            WriteSpan(MemoryMarshal.AsBytes(value));
+        }
+        #endregion
+
+        #region Writing VarLength types
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteVarInt<T>(T value) where T : Enum
+        {
+            WriteVarInt(Unsafe.As<T, int>(ref value));
+        }
+
+        public unsafe void WriteVarInt(int value)
+        {
+            var ptr = stackalloc byte[5];
+
+            var unsigned = (uint)value;
+            var size = 0;
+            do
+            {
+                var temp = (byte) (unsigned & 127);
+                unsigned >>= 7;
+
+                if (unsigned != 0)
+                    temp |= 128;
+
+                ptr[size++] = temp;
+            } while (unsigned != 0);
+            
+            fixed (byte* bptr = buffer)
+                System.Buffer.MemoryCopy(ptr, bptr, buffer.Length, size);
+            index += size;
```

Comment:
The old version had some optimizations done to it. I wonder what the difference in performance is here.